amaranth/hdl/ast.py

```diff
@@ -1929,6 +1929,8 @@ class ValueSet(_MappedKeySet):
 
 class SignalKey:
     def __init__(self, signal):
+        if isinstance(signal, ValueCastable):
+            signal = Value.cast(signal)
```

Comment:
Why do we need this?

Reply:
I'd say this is a consequence of RFC 15; a `Signal` constructed from a `ShapeCastable` becomes a `ValueCastable` around the signal.
When constructing a `SignalKey` from such a signal, in this case `port.data`, we need to unwrap the underlying signal.

Reply:
So this comes back to the additions to `hdl/xfrm.py`.